tendermint/src/lite_impl/signed_header.rs

```diff
@@ -52,68 +48,89 @@ impl lite::Commit for block::signed_header::SignedHeader {
     }
 
     fn validate(&self, vals: &Self::ValidatorSet) -> Result<(), Error> {
-        if self.commit.precommits.len() != vals.validators().len() {
+        if self.commit.signatures.len() != vals.validators().len() {
             fail!(
                 Kind::ImplementationSpecific,
                 "pre-commit length: {} doesn't match validator length: {}",
-                self.commit.precommits.len(),
+                self.commit.signatures.len(),
                 vals.validators().len()
             );
         }
 
-        for precommit_opt in self.commit.precommits.iter() {
-            match precommit_opt {
-                Some(precommit) => {
-                    // make sure each vote is for the correct header
-                    if let Some(header_hash) = precommit.header_hash() {
-                        if header_hash != self.header_hash() {
-                            fail!(
-                                Kind::ImplementationSpecific,
-                                "validator({}) voted for header {}, but current header is {}",
-                                precommit.validator_address,
-                                header_hash,
-                                self.header_hash()
-                            );
-                        }
-                    }
-
-                    // returns FaultyFullNode error if it detects a signer isn't present in the validator set
-                    if vals.validator(precommit.validator_address) == None {
-                        let reason = format!(
-                            "Found a faulty signer ({}) not present in the validator set ({})",
-                            precommit.validator_address,
-                            vals.hash()
-                        );
-                        fail!(Kind::FaultyFullNode, reason);
-                    }
+        for commit_sig in self.commit.signatures.iter() {
+            // returns FaultyFullNode error if it detects a signer that is not present in the validator set
+            if let Some(val_addr) = commit_sig.validator_address {
+                if vals.validator(val_addr) == None {
+                    let reason = format!(
+                        "Found a faulty signer ({}) not present in the validator set ({})",
+                        val_addr,
+                        vals.hash()
+                    );
+                    fail!(Kind::FaultyFullNode, reason);
```

Comment:
The var `reason` can be deleted and this can be simplified to:
```suggestion
                    fail!(
                        Kind::FaultyFullNode,
                        "Found a faulty signer ({}) not present in the validator set ({})",
                        val_addr,
                        vals.hash()
                    );
```